peekingduck/pipeline/nodes/heuristic/fps.py

```diff
@@ -0,0 +1,92 @@
+# Copyright 2021 AI Singapore
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Calculates the FPS of video
+"""
+
+from typing import Any, Dict, List
+from time import perf_counter
+import numpy as np
+
+from peekingduck.pipeline.nodes.node import AbstractNode
+
+NUM_FRAMES = 14
+
+
+class Node(AbstractNode):
+    """ FPS node class that calculates the FPS of the image frame """
+
+    def __init__(self, config: Dict[str, Any]) -> None:
+        super().__init__(config, node_path=__name__)
+        self.time_window = [float(0)]
+
+        self.moving_avg_thres = config["moving_avg"]
+        self.moving_average_fps: List[float] = []
+        self.prev_frame_timestamp = 0.0
+        self.global_avg_fps = 0.0
+        self.count = 0
+
+    def run(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
+        """ Calculates FPS using the time difference between the current
+        frame and the previous frame.
+
+        Args:
+            inputs: ["pipeline_end"]
+
+        Returns:
+            outputs (dict): Dict with key "fps".
+        """
+
+        if len(self.time_window) > NUM_FRAMES:
+            self.time_window.pop(0)
+
+        # "Dampened" FPS that reduces jitter effect of instantaneous FPS
+        curr_frame_timestamp = perf_counter()
+        self.time_window.append(curr_frame_timestamp)
+        num_frames = len(self.time_window)
+        time_diff = self.time_window[-1] - self.time_window[0]
+        average_fps = num_frames / time_diff
```

Comment:
There are 2 methods of getting the moving average FPS:
1. The above, where times are added to a list
2. Averaging the FPS per frame as per your _moving_average function

For approach 1 in the above chunk of code, the inherent flaw is that `self.time_window` will end up looking like: [0.0, 6.985116089, 8.039921208, 8.101609373] and the first value of zero affects calcs.

Suggest removing this, and using only approach 2 instead.